consensus/parlia/fusionfork.go

```diff
@@ -0,0 +1,226 @@
+	"container/heap"
+	"context"
+	"math"
+	"math/big"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/systemcontracts"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/internal/ethapi"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// init contract
+func (p *Parlia) initFusionContract(state *state.StateDB, header *types.Header, chain core.ChainContext,
+	txs *[]*types.Transaction, receipts *[]*types.Receipt, receivedTxs *[]*types.Transaction, usedGas *uint64, mining bool,
+) error {
+	// method
+	method := "initialize"
+	// contracts
+	contracts := []string{
+		systemcontracts.StakeHubContract,
+	}
+	// get packed data
+	data, err := p.stakeHubABI.Pack(method)
+	if err != nil {
+		log.Error("Unable to pack tx for init fusion contract", "error", err)
+		return err
+	}
+	for _, c := range contracts {
+		msg := p.getSystemMessage(header.Coinbase, common.HexToAddress(c), data, common.Big0)
+		// apply message
+		log.Trace("init fusion contract", "block hash", header.Hash(), "contract", c)
+		err = p.applyTransaction(msg, state, header, chain, txs, receipts, receivedTxs, usedGas, mining)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type ValidatorItem struct {
+	address     common.Address
+	votingPower *big.Int
+}
+
+// An ValidatorHeap is a min-heap of validator's votingPower.
+type ValidatorHeap []ValidatorItem
+
+func (h ValidatorHeap) Len() int { return len(h) }
+
+// We want topK validators with max voting power, so we need a min-heap
+func (h ValidatorHeap) Less(i, j int) bool {
+	if h[i].votingPower.Cmp(h[j].votingPower) == 0 {
+		return h[i].address.Hex() < h[j].address.Hex()
+	} else {
+		return h[i].votingPower.Cmp(h[j].votingPower) == -1
+	}
+}
+func (h ValidatorHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
+
+func (h *ValidatorHeap) Push(x interface{}) {
+	// Push and Pop use pointer receivers because they modify the slice's length,
+	// not just its contents.
+	*h = append(*h, x.(ValidatorItem))
+}
+
+func (h *ValidatorHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+func (p *Parlia) updateEligibleValidators(state *state.StateDB, header *types.Header, chain core.ChainContext,
+	txs *[]*types.Transaction, receipts *[]*types.Receipt, receivedTxs *[]*types.Transaction, usedGas *uint64, mining bool,
+) error {
+	// 1. get all validators and its voting power
+	method := "getValidatorWithVotingPower"
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel() // cancel when we are finished consuming integers
+
+	blockNr := rpc.BlockNumberOrHashWithHash(header.ParentHash, false)
+	toAddress := common.HexToAddress(systemcontracts.StakeHubContract)
+	gas := (hexutil.Uint64)(uint64(math.MaxUint64 / 2))
```

Comment:
As our gas is extremely large, we don't worry we can get all the results.

So I think we can get all information just by one call. 

But can you have a test that if we have 1K validator,  does one call can get that?

Reply:
Done